Pretty-printing using String regexes

garypgryan opened the pull request:
Pretty-printing using String regexes

cmoesel commented:
I have not reviewed the code yet, but I gave this a run.  First of all, congrats for getting it working!  It feels pretty seamless and doesn't seem to affect performance.  That's great!

I have a couple of questions based on my observation of it:

1. I got an error right off the bat, but I can't tell why.  When I run the CLI without your changes, I don't see anything that looks like an error between the first two log lines.  Here's the error:
![image](https://user-images.githubusercontent.com/2278253/52006758-e9f19d00-249a-11e9-8bad-446c8cde5671.png)

2. The first time we see a unique _WARN_ or _ERROR_ it is in more colorful text, but subsequent times (and for _INFO_), it is red text.  Is this what @markkramerus was looking for?  It seems like if we want to de-emphasize the "duplicated" logs, we should pick a color other than red.  Red makes my brain think "bad".

3. Is there a mode to turn off the messages that are detected as "duplicates"?  If so, we should document it.  If not, that's fine.

4. It seems that the duplicate detection strategy is based on error number alone.  I know that long term we want to be smarter about it.  I just want to confirm with @markkramerus that it is OK to releasing this interim approach into master.

5. Why is the _Suggested Fix_ always wrapped in single-quotes?  I think it would look more natural just to put the message without quotes.  It especially looks weird when the fix is unknown:
![image](https://user-images.githubusercontent.com/2278253/52007277-2376d800-249c-11e9-8880-e066f72ad8a3.png)

6. There are some messages that seem to have a leading space in them.  For example:
![image](https://user-images.githubusercontent.com/2278253/52007323-3be6f280-249c-11e9-92b8-7de843e76507.png)

cmoesel commented:
Also, I rebased this on the current master branch in order to avoid the conflicts.  Let me know if you'd like me to force-push the rebased branch.  It will require you to reset your branch locally.  Otherwise we can wait until the PR is ready to be merged and just rebase it again then.

cmoesel commented on runCli.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+node . ../shr_spec-master/spec -l error
```
I don't think this should be checked in, especially since others probably don't have a `shr_spec-master` folder.

garypgryan replied:
removed runCli.sh

garypgryan commented:
Thanks for the tips; they are most helpful.    I'm getting acquainted with
ES6 idioms and best practices so your explanations of why you would change
things is instructive.    I don't think it will take a lot to implement the
changes.

Gary

On Wed, Jan 30, 2019 at 4:11 PM Chris Moesel <notifications@github.com>
wrote:

> *@cmoesel* requested changes on this pull request.
>
> Overall, looks and works well! See my prior overall comment on the PR, but
> also my individual comments on specific lines of code.
>
> I know ES6 is new to you, so I also tried to provide some pointers on best
> practices, etc.
>
> I also know you already have the "next generation" started, so if any of
> this will require significant effort, we may want to discuss if it's worth
> the effort when we know it will be replaced soon.
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252412467>
> :
>
> > @@ -0,0 +1,162 @@
> +const Transform = require('stream').Transform;
> +var fs = require('fs');
>
> In Node.js and ES6 we generally avoid var because it has weird scoping
> rules and is automatically "hoisted" to the top of the current scope. I
> find this confusing (as do others). It's better just to use const and let.
>
> In general, when declaring a variable:
>
>    - use const if it will never be re-assigned
>    - use let if it will be re-assigned (or the value isn't known when
>    it's declared)
>    - never use var unless you absolutely know *why* you need it (I
>    haven't found that case yet)
>
> I just found this post which I generally agree with:
> https://medium.com/javascript-scene/javascript-es6-var-let-or-const-ba58b8dcde75
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252412944>
> :
>
> > @@ -0,0 +1,162 @@
> +const Transform = require('stream').Transform;
> +var fs = require('fs');
>
> BTW, in this specific case, it should be const fs. We pretty much *always*
> use const with require.
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252415596>
> :
>
> > +const chalk = require('chalk');   //library for colorizing Strings
> +// color palette for messages -- can also do rgb; e.g.  chalk.rgb(123, 45, 67)
> +const originalErrorColor = chalk.bold.redBright;
> +const errorDetailColor = chalk.bold.cyan;
> +const errorCodeColor = chalk.bold.greenBright;
> +
> +//https://stackoverflow.com/questions/48507828/pipe-issue-with-node-js-duplex-stream-example
> +
> +// make a class that implements a Duplex stream; this means you can use it to pipe into and out of
> +class PrettyPrintDuplexStream extends Transform {
> +
> +  constructor(name, options) {
> +    super(options);
> +    this.name = name;
> +    this.solutionMap = {};
> +    let csvFilePath = 'errorMessages.txt';
>
> Specifying the file path like this will expect the errorMessages.txt file
> to be in the *working directory*. This will cause problems if you try to
> invoke the CLI from another directory. That's probably rare, but we should
> still try to handle the case.
>
> The best way to do that is to construct the path based on it's relation to *this
> file* (*PrettyPrintDuplexStream.js*). In Node.js, there is a special
> variable __dirname (note: *two* leading underscores) that represents the
> path of the directory in which the *current file* resides. There is also
> a library called path that helps deal with file paths in an
> OS-independent way (since Windows uses \ and everyone else uses /). So...
> putting those together:
>
> const path = require('path'); // put this at top of file// ... all those other lines of code up until this one, then:const csvFilePath = path.join(__dirname, 'errorMessages.txt');
>
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252416342>
> :
>
> > @@ -0,0 +1,162 @@
> +const Transform = require('stream').Transform;
> +var fs = require('fs');
>
> And I won't comment on any of the other var and let uses, but please fix
> them all.
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252416785>
> :
>
> > +// color palette for messages -- can also do rgb; e.g.  chalk.rgb(123, 45, 67)
> +const originalErrorColor = chalk.bold.redBright;
> +const errorDetailColor = chalk.bold.cyan;
> +const errorCodeColor = chalk.bold.greenBright;
> +
> +//https://stackoverflow.com/questions/48507828/pipe-issue-with-node-js-duplex-stream-example
> +
> +// make a class that implements a Duplex stream; this means you can use it to pipe into and out of
> +class PrettyPrintDuplexStream extends Transform {
> +
> +  constructor(name, options) {
> +    super(options);
> +    this.name = name;
> +    this.solutionMap = {};
> +    let csvFilePath = 'errorMessages.txt';
> +    var array = fs.readFileSync(csvFilePath).toString().split('\n');
>
> Not to be nit-picky, but I think a more descriptive name would be good.
> For example, lines would be good, since this represents lines from the
> file.
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252417429>
> :
>
> > +
> +  constructor(name, options) {
> +    super(options);
> +    this.name = name;
> +    this.solutionMap = {};
> +    let csvFilePath = 'errorMessages.txt';
> +    var array = fs.readFileSync(csvFilePath).toString().split('\n');
> +    this.eCode = '-1';   // put in a value for error messages that don't have an ERROR_CODE
> +    this.solutionMap[this.eCode] = 'Error message has no error code';
> +
> +    // populate a map with the key as the ERROR_CODE number and the value is the suggested solution
> +    for( var i in array) {
> +      let line = array[i].toString();
> +      let parts= line.split(',');
> +      let key = parts[0].trim();
> +      let value = parts[2];
>
> This is fine, but it seems it probably could have been expressed more
> succinctly using a regular expression with some captures.
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252418356>
> :
>
> > +    this.solutionMap = {};
> +    let csvFilePath = 'errorMessages.txt';
> +    var array = fs.readFileSync(csvFilePath).toString().split('\n');
> +    this.eCode = '-1';   // put in a value for error messages that don't have an ERROR_CODE
> +    this.solutionMap[this.eCode] = 'Error message has no error code';
> +
> +    // populate a map with the key as the ERROR_CODE number and the value is the suggested solution
> +    for( var i in array) {
> +      let line = array[i].toString();
> +      let parts= line.split(',');
> +      let key = parts[0].trim();
> +      let value = parts[2];
> +      this.solutionMap[key] = value;
> +    }
> +    this.idSet = new Set();
> +    this.idSet.add(-1);
>
> What is this line for? If it's intended to add the code for "no error
> code", then should it be '-1' (as a string rather than an integer)?
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252420633>
> :
>
> > +    }
> +    this.idSet = new Set();
> +    this.idSet.add(-1);
> +  }
> +
> +  translateNames( inName ) {   // translate error Strings to friendlier informative alternatives.
> +    if (inName === 'shr-expand') {
> +      return('FHIR Mapping(expansion)');
> +    }
> +    if (inName === 'shr-fhir-export') {
> +      return('FHIR Export');
> +    }
> +    if (inName === 'shr-expand') {
> +      return('FHIR Mapping(expansion)');
> +    }
> +    return(inName);
>
> You have shr-expand twice above, and the friendlier string for it isn't
> quite right. Also, this is nit-picky (and totally a matter of preference),
> but it seems like a great opportunity for a switch statement! See below:
>
> switch (inName) {case 'shr-expand': return('Model Expansion');case 'shr-fhir-export': return('FHIR Export');default: return inName;
> }
>
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252422943>
> :
>
> > +
> +  getUnqualifiedName ( inName ) {     // take a name with '.' delimiters and return the last part
> +    let nameParts = inName.split('.');
> +    if (nameParts.length > 0) {
> +      return(nameParts[nameParts.length -1 ]);
> +    }
> +    return(inName);
> +  }
> +
> +  getMatchWithRegexPos(myStr, myRegex, pos) { // apply regex to String; return match at pos if it matches; otherwise the original string
> +    let myMatch  = myStr.match(myRegex);
> +    let myMsg = '';
> +    if (myMatch != null) {
> +      myMsg = myMatch[pos];
> +    }
> +    return(myMsg);
>
> Some people hate ternary operators, but if you don't hate them, then you
> could replace the 5 lines above (59-63) with:
>
> return myMatch != null ? myMatch[pos] : '';
>
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252427830>
> :
>
> > +
> +  parseFromPrefixOrSuffix( myRegex, myPrefix, mySuffix ) {
> +    let myPart = this.getMatchWithRegexPos(mySuffix, myRegex, 1);  //parse field from suffix
> +    if (myPart === '') {                                      // if missing, try prefix
> +      myPart = this.getMatchWithRegexPos(myPrefix, myRegex, 1);
> +    }
> +    return(myPart);
> +  }
> +
> +  // this function processes a single error message and returns a colorized, formatted string
> +  // for the moment, it writes the original input message in red to console.log
> +  processLine(myinline, printAllErrors) {
> +
> +    let myline = myinline.toString();
> +    let reg = /ERROR_CODE:([\d]+)\s/;   //match ERROR_CODE:ddddd
> +    let result = myline.match(reg);
>
> You don't really need the intermediately declared reg variable. I'd just
> use:
>
> let result = myline.match( /ERROR_CODE:([\d]+)\s/);
>
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252428958>
> :
>
> > +const originalErrorColor = chalk.bold.redBright;
> +const errorDetailColor = chalk.bold.cyan;
> +const errorCodeColor = chalk.bold.greenBright;
> +
> +//https://stackoverflow.com/questions/48507828/pipe-issue-with-node-js-duplex-stream-example
> +
> +// make a class that implements a Duplex stream; this means you can use it to pipe into and out of
> +class PrettyPrintDuplexStream extends Transform {
> +
> +  constructor(name, options) {
> +    super(options);
> +    this.name = name;
> +    this.solutionMap = {};
> +    let csvFilePath = 'errorMessages.txt';
> +    var array = fs.readFileSync(csvFilePath).toString().split('\n');
> +    this.eCode = '-1';   // put in a value for error messages that don't have an ERROR_CODE
>
> I don't understand why eCode is a class-level variable. It seems to me
> like it should just be a local variable in the processLine function.
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252429373>
> :
>
> > +    if (myPart === '') {                                      // if missing, try prefix
> +      myPart = this.getMatchWithRegexPos(myPrefix, myRegex, 1);
> +    }
> +    return(myPart);
> +  }
> +
> +  // this function processes a single error message and returns a colorized, formatted string
> +  // for the moment, it writes the original input message in red to console.log
> +  processLine(myinline, printAllErrors) {
> +
> +    let myline = myinline.toString();
> +    let reg = /ERROR_CODE:([\d]+)\s/;   //match ERROR_CODE:ddddd
> +    let result = myline.match(reg);
> +    let preErrCode = '';
> +    let postErrCode = '';
> +    this.eCode = '-1';
>
> Again, it seems to me like this would be better as a local var (e.g., let
> eCode) instead of a class-level variable (e.g., this.eCode).
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252430093>
> :
>
> > +  // for the moment, it writes the original input message in red to console.log
> +  processLine(myinline, printAllErrors) {
> +
> +    let myline = myinline.toString();
> +    let reg = /ERROR_CODE:([\d]+)\s/;   //match ERROR_CODE:ddddd
> +    let result = myline.match(reg);
> +    let preErrCode = '';
> +    let postErrCode = '';
> +    this.eCode = '-1';
> +    if (printAllErrors) {
> +      console.log( originalErrorColor (myline.trim() ));   // print the input message in red (for now)
> +    }
> +    let formattedOutput = '\nERROR ';
> +
> +    // split myline on ERROR_CODE; preErrCode is everything before ERROR_CODE; postErrCode is everything after
> +    if (result !== undefined && result != null) {
>
> In JavaScript, if you use != null then it will be false for both null and
> undefined. On other words, you can replace the line above with just:
>
> if (result != null) {
>
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252430865>
> :
>
> > +    let reg = /ERROR_CODE:([\d]+)\s/;   //match ERROR_CODE:ddddd
> +    let result = myline.match(reg);
> +    let preErrCode = '';
> +    let postErrCode = '';
> +    this.eCode = '-1';
> +    if (printAllErrors) {
> +      console.log( originalErrorColor (myline.trim() ));   // print the input message in red (for now)
> +    }
> +    let formattedOutput = '\nERROR ';
> +
> +    // split myline on ERROR_CODE; preErrCode is everything before ERROR_CODE; postErrCode is everything after
> +    if (result !== undefined && result != null) {
> +      let temp = myline.split('ERROR_CODE');
> +      preErrCode = temp[0];
> +      postErrCode = temp[1];
> +      if ( result[1] === undefined || result[1] === null) {
>
> Again, this can be:
>
> if ( result[1] == null) {
>
> Note that is a ==, which is less strict than ===. If you use ===, then it
> must be an exact match (e.g., === null means it *must* be null). If you
> use ==, it uses looser matching semantics, so it can be null or undefined.
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252431895>
> :
>
> > +    this.eCode = '-1';
> +    if (printAllErrors) {
> +      console.log( originalErrorColor (myline.trim() ));   // print the input message in red (for now)
> +    }
> +    let formattedOutput = '\nERROR ';
> +
> +    // split myline on ERROR_CODE; preErrCode is everything before ERROR_CODE; postErrCode is everything after
> +    if (result !== undefined && result != null) {
> +      let temp = myline.split('ERROR_CODE');
> +      preErrCode = temp[0];
> +      postErrCode = temp[1];
> +      if ( result[1] === undefined || result[1] === null) {
> +        this.eCode = '-1';
> +      }
> +      else {
> +        this.eCode = result[1].trim();   //gpg
>
> //gpg?
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252433723>
> :
>
> > +      postErrCode = temp[1];
> +      if ( result[1] === undefined || result[1] === null) {
> +        this.eCode = '-1';
> +      }
> +      else {
> +        this.eCode = result[1].trim();   //gpg
> +      }
> +    }
> +    let dateTimeRegex = /\[\d\d:\d\d:\d\d.\d\d\dZ\]\s+/;
> +    let outline  = myline.replace(dateTimeRegex,'').toString();  //remove timestamp; format is [hh.mm.ss.xxxZ]
> +    let errShrRegex = /(ERROR[\s]+[\w]+:[\s]+)/;
> +    // split into piece before "ERROR_CODE" and piece after "ERROR_CODE"
> +    preErrCode = preErrCode.replace(errShrRegex, '');    // remove the 'ERROR shr' part
> +    preErrCode = preErrCode.replace(dateTimeRegex, ''); // remove the timestamp
> +    let detailMsg = preErrCode;
> +    formattedOutput += this.eCode + ': ' + detailMsg;  // first part of new message is ERROR xxxxx: <<detail>>
>
> FYI, not a big deal, but just so you know -- JavaScript has string
> interpolation, so you could do:
>
> formattedOutput += `${this.eCode}: ${detailMsg}`;
>
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252434675>
> :
>
> > +
> +    if (this.idSet.has(key)) {
> +      return('');
> +    }
> +    else {
> +      this.idSet.add(key);
> +      if (printAllErrors === false) {
> +        console.log( originalErrorColor (myline ));
> +      }
> +      return (outline);
> +    }
> +  }
> +
> +
> +  _transform(chunk, encoding, callback) {
> +    //this.push(chunk);
>
> Please remove this commented out code.
> ------------------------------
>
> In PrettyPrintDuplexStream.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252435220>
> :
>
> > +    +   errorDetailColor ( '\n    During:         ' + this.translateNames(modulePart))
> +    +   errorDetailColor( '\n    Class:          ' + this.getUnqualifiedName(this.translateNames(shrIdPart)));
> +
> +    // if parts are optional/missing, then only print them if they are found
> +    if (targetSpecPart != '') {
> +      outline += errorDetailColor( '\n    Target Spec:    ' + this.translateNames(targetSpecPart));
> +    }
> +    if (targetPart != '') {
> +      outline += errorDetailColor( '\n    Target Class:   ' + this.translateNames(targetPart));
> +    }
> +    if (mappingRulePart != '') {
> +      outline += errorDetailColor( '\n    Mapping Rule:   ' + this.translateNames(mappingRulePart)) ;
> +    }
> +
> +    // lookup the suggested fix using eCode as the key
> +    outline += errorDetailColor( '\n    Suggested Fix:  ' + this.solutionMap[this.eCode]).trim() + '\n'  ;
>
> I wonder if we should only print out the *Suggested Fix* if we have one?
> I don't think there's a lot of value in having:
>
>     Suggested Fix:  Unknown
>
> It seems like it might be better just not to print that line at all.
> ------------------------------
>
> In app.js
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252435388>
> :
>
> >  // Set up the logger streams
>  const [ll, lm] = [program.logLevel.toLowerCase(), program.logMode.toLowerCase()];
>  const streams = [];
>  if (lm == 'short' || lm == 'long') {
>    const prettyStdOut = new bps({mode: lm});
> -  prettyStdOut.pipe(process.stdout);
> +  prettyStdOut.pipe(mdpStream);
> +  mdpStream.pipe(process.stdout);
>
> It's pretty cool how this works. I like it.
> ------------------------------
>
> In runCli.sh
> <https://github.com/standardhealth/shr-cli/pull/148#discussion_r252436335>
> :
>
> > @@ -0,0 +1,2 @@
> +#!/bin/bash
> +node . ../shr_spec-master/spec -l error
>
> I don't think this should be checked in, especially since others probably
> don't have a shr_spec-master folder.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/standardhealth/shr-cli/pull/148#pullrequestreview-198274433>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ae9Kdzli8g8y58eFee64bhwKLQd_u5A0ks5vIgqXgaJpZM4aWmaS>
> .
>

garypgryan commented on PrettyPrintDuplexStream.js:
```diff
@@ -0,0 +1,162 @@
+    this.name = name;
+    this.solutionMap = {};
+    let csvFilePath = 'errorMessages.txt';
+    var array = fs.readFileSync(csvFilePath).toString().split('\n');
```
changed to allLines.
const allLines = fs.readFileSync(csvFilePath).toString().split('\n');

garypgryan commented on PrettyPrintDuplexStream.js:
```diff
@@ -0,0 +1,162 @@
+    super(options);
+    this.name = name;
+    this.solutionMap = {};
+    let csvFilePath = 'errorMessages.txt';
```
Changed to 
const csvFilePath = path.join(__dirname, 'errorMessages.txt' );

garypgryan commented on PrettyPrintDuplexStream.js:
```diff
@@ -0,0 +1,162 @@
+const Transform = require('stream').Transform;
+var fs = require('fs');
```
done

garypgryan commented on PrettyPrintDuplexStream.js:
```diff
@@ -0,0 +1,162 @@
+      this.solutionMap[key] = value;
+    }
+    this.idSet = new Set();
+    this.idSet.add(-1);
```
changed to 
    this.idSet = new Set();
    this.idSet.add( noErrorCode );

garypgryan commented on PrettyPrintDuplexStream.js:
```diff
@@ -0,0 +1,162 @@
+    preErrCode = preErrCode.replace(errShrRegex, '');    // remove the 'ERROR shr' part
+    preErrCode = preErrCode.replace(dateTimeRegex, ''); // remove the timestamp
+    let detailMsg = preErrCode;
+    formattedOutput += this.eCode + ': ' + detailMsg;  // first part of new message is ERROR xxxxx: <<detail>>
```
Changed to 
formattedOutput += `${eCode}: ${preErrCode}`;  // first part of new message is ERROR xxxxx: <<detail>>

garypgryan commented on PrettyPrintDuplexStream.js:
```diff
@@ -0,0 +1,162 @@
+
+  
+  _transform(chunk, encoding, callback) {
+    //this.push(chunk);
```
done.

garypgryan commented on PrettyPrintDuplexStream.js:
```diff
@@ -0,0 +1,162 @@
+    }
+  
+    // lookup the suggested fix using eCode as the key
+    outline += errorDetailColor( '\n    Suggested Fix:  ' + this.solutionMap[this.eCode]).trim() + '\n'  ;
```
done.

garypgryan commented on app.js:
```diff
@@ -70,12 +70,17 @@ const importCimcore = program.importCimcore;
   const prettyStdOut = new bps({mode: lm});
-  prettyStdOut.pipe(process.stdout);
+  prettyStdOut.pipe(mdpStream);
+  mdpStream.pipe(process.stdout);
```
Thanks!

garypgryan commented on PrettyPrintDuplexStream.js:
```diff
@@ -0,0 +1,162 @@
+      let line = array[i].toString();
+      let parts= line.split(',');
+      let key = parts[0].trim();
+      let value = parts[2];
```
I want to rewrite this code so that the resource file can support comments (i.e. lines that start with // ) and also embedded commas in the error message strings (which this code doesn't support at the moment).

ngfreiter commented:
This PR has been replaced by #227.
